python/pyiceberg/expressions/base.py

```diff
@@ -672,3 +683,42 @@ def visit_unbound_predicate(self, predicate) -> BooleanExpression:
 
     def visit_bound_predicate(self, predicate) -> BooleanExpression:
         raise TypeError(f"Found already bound predicate: {predicate}")
+
+
+class BoundBooleanExpressionVisitor(BooleanExpressionVisitor[T], ABC):
+    @abstractmethod
+    def visit_in(self, ref: BoundReference[T], literals: set[Literal[Any]]) -> T:
+        """Visit the ref and literals from an In boolean expression
+        Args:
+            ref (BoundReference[T]): The ref used in the In boolean expression
+            literals (Tuple[Literal[T], ...]): The literals used in the In boolean expression
+        Raises:
+            NotImplementedError: If the concrete bound boolean expression visitor does not override this method
+        Returns:
+            R: The return type defined by the concrete bound boolean expression visitor
+        """
+
+    def visit_unbound_predicate(self, predicate: UnboundPredicate[T]):
+        """Visit an unbound predicate
+        Args:
+            predicate (UnboundPredicate[T]): An unbound predicate
+        Raises:
+            TypeError: This always raises since an unbound predicate is not expected in a bound boolean expression
+        """
+        raise TypeError(f"Not a bound predicate: {predicate}")
+
+    def visit_bound_predicate(self, predicate: BoundPredicate[T]) -> T:
+        """Visit a bound predicate
+        Args:
+            predicate (BoundPredicate[T]): A bound predicate
+        Raises:
+            TypeError: When an unsupported predicate is found
+        Returns:
+            R: The return type defined by the concrete bound boolean expression visitor
+        """
+        if not isinstance(predicate.term, BoundReference):
+            raise TypeError(f"{type(self).__name__} does not support non-reference: {predicate.term}")
```

Comment:
I don't think that we need to fail here. This is most likely an artifact of not having `Transform` expressions at first in the Java implementation.